internal/management/controller/instance_controller.go

```diff
@@ -1114,7 +1113,7 @@ func (r *InstanceReconciler) reconcilePrimary(ctx context.Context, cluster *apiv
 		if err := r.handlePromotion(ctx, cluster); err != nil {
 			return false, err
 		}
-		restarted = true
+		promoted = true
```

Comment:
This is not your PR but the existing code.
This is asymmetric. We decide that the instance is a primary with IsPrimary, which checks for the existence of "recovery.conf"

Let's say we:
1. successfully promote the instace
2. when trying to update the Status or drop the connections, we fail

Then the next time around we'll see the instance is a primary, and we won't retry.

Reply:
If we don't update the status, the Status patch step is retried.
The drop connection is not vital. It is there only to prevent a long-running session through the read-only service from silently becoming a read-write one.